File: 1. 단계별로_풀어보기/20~29/27. 그래프와 순회/5. 2606.py
from collections import deque
import sys

input = sys.stdin.readline
output = sys.stdout.write

N = int(input())
M = int(input())
pairArr = list(list(map(int, input().split())) for i in range(M))

connArr = list(deque() for i in range(N + 1))
for a, b in pairArr:
  connArr[a].append(b)
  connArr[b].append(a)

isInfected = [False] * (N + 1)

nextInfect = deque()

def bfs(now):
  for next in connArr[now]:
    if isInfected[next] == False:
      isInfected[next] = True
      nextInfect.append(next)

  if len(nextInfect) > 0:
    next = nextInfect.popleft()
    bfs(next)

isInfected[1] = True
bfs(1)

output(f"{sum(1 for val in isInfected if val == True) - 1}")

File: 1. 단계별로_풀어보기/20~29/27. 그래프와 순회/5. 2606.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const N = +inputArr[0]
const M = +inputArr[1]
const pairArr = inputArr.slice(2).map(row => row.trim().split(' ').map(Number))

// 연결 정보 저장
const mapArr = Array.from({length: N + 1}, () => {
  return {}
})
pairArr.forEach(ab => {
  const [a, b] = ab
  mapArr[a][b] = 1
  mapArr[b][a] = 1
})

const connArr = Array.from({length: N + 1}, () => [])
mapArr.forEach((maps, mapIdx) => {
  connArr[mapIdx] = Object.keys(maps)
})

// 감염여부 저장 배열
const isInfected = Array.from({length: N + 1}).fill(false)

isInfected[1] = true

// 감염 시작
dfs(1)

// 결과 변수
let result = isInfected.filter(val => val).length - 1

// 출력
process.stdout.write(`${result}`)

function dfs(now) {
  connArr[now].forEach(next => {
    if (isInfected[next] === false) {
      isInfected[next] = true
      dfs(next)
    }
  })
}
